modules/scheduler-impl/src/main/java/org/opencastproject/scheduler/impl/persistence/SchedulerServiceDatabaseImpl.java

```diff
@@ -169,40 +159,79 @@ public Date getLastModified(String agentId) throws NotFoundException, SchedulerS
   }
 
   @Override
-  public String getTransactionId(String source) throws NotFoundException, SchedulerServiceDatabaseException {
-    EntityManager em = null;
-    try {
-      em = emf.createEntityManager();
-      String orgId = securityService.getOrganization().getId();
-      Query q = em.createNamedQuery("Transaction.findBySource").setParameter("source", source).setParameter("org",
-              orgId);
-      TransactionDto dto = (TransactionDto) q.getSingleResult();
-      return dto.getId();
-    } catch (NoResultException e) {
-      throw new NotFoundException("Transaction with source " + source + " does not exist");
-    } catch (Exception e) {
-      logger.error("Could not retrieve id for transaction with source '{}': {}", source, getStackTrace(e));
-      throw new SchedulerServiceDatabaseException(e);
-    } finally {
-      if (em != null)
-        em.close();
-    }
-  }
-
-  @Override
-  public String getTransactionSource(String id) throws NotFoundException, SchedulerServiceDatabaseException {
+  public void storeEvent(String mediapackageId, String organizationId, Opt<String> captureAgentId, Opt<Date> start,
+          Opt<Date> end, Opt<String> source, Opt<String> recordingState, Opt<Long> recordingLastHeard,
+          Opt<String> reviewStatus, Opt<Date> reviewDate, Opt<String> presenters, Opt<Boolean> optOut,
+          Opt<String> lastModifiedOrigin, Opt<Date> lastModifiedDate, Opt<String> checksum,
+          Opt<Map<String, String>> workflowProperties, Opt<Map<String, String>> caProperties
+  ) throws SchedulerServiceDatabaseException {
     EntityManager em = null;
+    EntityTransaction tx = null;
     try {
       em = emf.createEntityManager();
-      Opt<TransactionDto> entity = getTransactionDto(id, em);
-      if (entity.isNone())
-        throw new NotFoundException("Transaction with ID " + id + " does not exist");
+      tx = em.getTransaction();
+      tx.begin();
+      Opt<ExtendedEventDto> entityOpt = getExtendedEventDto(mediapackageId, organizationId, em);
+      ExtendedEventDto entity = entityOpt.getOr(new ExtendedEventDto());
+      entity.setMediaPackageId(mediapackageId);
+      entity.setOrganization(organizationId);
+      if (captureAgentId.isSome()) {
+        entity.setCaptureAgentId(captureAgentId.get());
+      }
+      if (start.isSome()) {
```

Comment:
Just wondering but is the start and end date really optional?
I expect things to blow up somewhere if there is none ;-)

Reply:
Well, to be honest, start and end are not optional, but the method is used to persist new events as well as to update them. So not providing start and end date only makes sense when updating some other fields of an existing event. This is not a good design, I admit. But since this is an internal class which is used by the scheduler implementation only, I think it doesn't make sense to be extra cautious and add a whole bunch of checks here.